File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.cpp
class Solution {
public:
    void dfs(vector <vector <int>> &grid, int r, int c, int &ans, int &tot) {
        ans = max(ans, tot);
        if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == 0) return;
        tot += grid[r][c];
        grid[r][c] = 0;
        dfs(grid, r + 1, c, ans, tot);
        dfs(grid, r - 1, c, ans, tot);
        dfs(grid, r, c + 1, ans, tot);
        dfs(grid, r, c - 1, ans, tot);
    }

    int findMaxFish(vector<vector<int>>& grid) {
        if (grid.size() == 0) return 0;
        int ans = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] > 0) {
                    int tot = 0;
                    dfs(grid, i, j, ans, tot);
                }
            }
        }
        return ans;
    }
};

File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.go
func findMaxFish(grid [][]int) int {
    m, n := len(grid), len(grid[0])

    vis := make([][]bool, m)
    for i := range vis {
        vis[i] = make([]bool, n)
    }
    
    var ds = []struct{dx, dy int}{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

    var dfs func(r, c int) int
    dfs = func(r, c int) int {
        if r < 0 || c < 0 || r >= m || c >= n || vis[r][c] || grid[r][c] == 0 {
            return 0
        }
        vis[r][c] = true
        fc := grid[r][c]
        for _, dir := range ds {
            fc += dfs(r + dir.dx, c + dir.dy)
        }
        return fc
    }

    ans := 0
    
    for r := 0; r < m; r++ {
        for c := 0; c < n; c++ {
            if grid[r][c] > 0 && !vis[r][c] {
                ans = max(ans, dfs(r, c))
            }
        }
    }
    
    return ans
}